gen/field.go

```diff
@@ -183,15 +183,23 @@ func compileJSONTag(f *compile.FieldSpec, name string, opts ...string) *structta
 	// defined as false, 0, a nil pointer, a nil interface value, and any empty
 	// array, slice, map, or string.
 	//
-	// If the field is marked with "!omitempty", then "omitempty" will not be added
-	// For example:
-	//
-	// struct MyStruct {
-	//   1: optional list<Option> options  (go.tag='json:"options,!omitempty"')
-	// }
+	// If the field is marked with "!omitempty", then "omitempty" will not be added.
+	// If both "!omitempty" and "omitempty" are present, then "omitempty" is removed.
+	hasNotOmitempty := t.HasOption(notOmitempty)
+	hasOmitempty := t.HasOption(omitempty)
 	if (isReferenceType(f.Type) || isStructType(f.Type) || isPrimitiveType(f.Type)) &&
-		!f.Required && !t.HasOption(notOmitempty) && !t.HasOption(omitempty) {
+		!f.Required && !hasNotOmitempty && !hasOmitempty {
 		t.Options = append(t.Options, omitempty)
+	} else if hasNotOmitempty && hasOmitempty {
```

Comment:
I believe these omitempty conditionals should go inside a nested if clause after it has determined that omitempty would be needed on a field. That is, up to:

```
if (isReferenceType(f.Type) || isStructType(f.Type) || isPrimitiveType(f.Type)) && !f.isRequired {
  // We would normally add omitempty here.
}
```
Inside that block, we can add the logic of adding it, removing it, or leaving it.